### PR TITLE
Port the particle partition in buffers to GPU

### PR DESCRIPTION
**Pull request by RemiLehe**
This PR ports the particle partition to GPU.

# Implementation

I added a few helper function that hide away the details of how this is done on CPU and GPU respectively. These helper function are in a new file `SortingUtils.H`. @atmyers Should some of these function be replaced by `amrex` equivalents?

# Performance tests

In a simple 2D test (see figure below), the performance numbers confirm that `partition` now fully runs on GPU:

- With the `dev` branch:
```
---------------------------------------------------------------------------------------------
Name                                          NCalls  Excl. Min  Excl. Avg  Excl. Max   Max %
---------------------------------------------------------------------------------------------
PPC::Evolve::partition                          2256      5.535      5.535      5.535   8.59%
```
```
TinyProfiler total time across processes [min...avg...max]: 64.4 ... 64.4 ... 64.4
```

- With this PR:
```
---------------------------------------------------------------------------------------------
Name                                          NCalls  Excl. Min  Excl. Avg  Excl. Max   Max %
---------------------------------------------------------------------------------------------
PPC::Evolve::partition                          2256     0.3138     0.3138     0.3138   0.52%
```
```
TinyProfiler total time across processes [min...avg...max]: 60.14 ... 60.14 ... 60.14
```

i.e. the current PR is faster because the `partition` now takes negligible time (0.52%)

# Accuracy tests

I also checked that the results looked correct, in the case where current deposition buffers are used:
Screenshot from 2019-09-30 15-04-50
Here, the current `Jz` in the patch is lower at the edges, because the particles located there are in the current buffers and deposit on the mother grid below.

When zooming in on the patch, the `dev` branch and this PR show the same results:
Screenshot from 2019-09-30 15-05-43

### CODE REVIEW

File: Source/Particles/Sorting/SortingUtils.H [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+            m_start_index = start_index;
+        };
+
+
+        AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
+        void operator()( const long i ) const {
+            // Select a particle
+            auto& p = m_particles[i+m_start_index];
+            // Find the index of the cell where this particle is located
+            amrex::IntVect iv;
```
Comment (atmyers) [inline Source/Particles/Sorting/SortingUtils.H]: You can use `amrex::getParticleCell`, in `AMReX_ParticleUtil.H`, here if you want.

File: Source/Particles/Sorting/SortingUtils.H [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+ * \param[inout] v Vector of integers, to be filled by this routine
+ */
+void fillWithConsecutiveIntegers( amrex::Gpu::DeviceVector<long>& v )
+{
+#ifdef AMREX_USE_GPU
+    // On GPU: Use thrust
+    thrust::sequence( v.begin(), v.end() );
+#else
+    // On CPU: Use std library
+    std::iota( v.begin(), v.end(), 0L );
```
Comment (atmyers) [inline Source/Particles/Sorting/SortingUtils.H]: We don't have wrappers for these at the moment in AMReX, so I think it is fine and good to include them here.

File: Source/Particles/Sorting/Partition.cpp [hunk 3]
```diff
@@ -41,56 +42,64 @@ PhysicalParticleContainer::PartitionParticlesInBuffers(
...
+    amrex::ParallelFor( np, fillBufferFlag(pti, bmasks, inexflag, Geom(lev), 0) );
+    // - Find the indices that reorder particles so that the last particles
+    //   are in the larger buffer
+    fillWithConsecutiveIntegers( pid );
+    auto sep = stablePartition( pid.begin(), pid.end(), inexflag );
+    // At the end of this step, `pid` contains the indices that should be used to
+    // reorder the particles, and `sep` is the position in the array that
+    // separates the particles that deposit/gather on the fine patch (first part)
+    // and the particles that deposit/gather in the buffers (last part)
+    long n_fine = iteratorDistance(pid.begin(), sep);
```
Comment (ax3l) [inline Source/Particles/Sorting/Partition.cpp]: maybe make this `const`?
```suggestion
    long const n_fine = iteratorDistance(pid.begin(), sep);
```

File: Source/Particles/Sorting/Partition.cpp [hunk 4]
```diff
@@ -41,56 +42,64 @@ PhysicalParticleContainer::PartitionParticlesInBuffers(
...
-            for (auto it = sep; it != pid.end(); ++it) {
-                const long id = *it;
-                const IntVect& iv = Index(aos[id], lev);
-                inexflag[id] = msk2(iv);
-            }
+            // - For each particle in the large buffer, find whether it is in
+            // the smaller buffer, by looking up the mask. Store the answer in `inexflag`.
+            amrex::ParallelFor( np - n_fine,
+               fillBufferFlag(pti, bmasks, inexflag, Geom(lev), n_fine) );
+            auto sep2 = stablePartition( sep, pid.end(), inexflag );
```
Comment (ax3l) [inline Source/Particles/Sorting/Partition.cpp]:
```suggestion
            auto const sep2 = stablePartition( sep, pid.end(), inexflag );
```

File: Source/Particles/Sorting/Partition.cpp [hunk 5]
```diff
@@ -41,56 +42,64 @@ PhysicalParticleContainer::PartitionParticlesInBuffers(
...
-    std::iota(pid.begin(), pid.end(), 0L);
-    auto sep = std::stable_partition(pid.begin(), pid.end(),
-                                     [&inexflag](long id) { return inexflag[id]; });
+    // - For each particle, find whether it is in the larger buffer,
+    //   by looking up the mask. Store the answer in `inexflag`.
+    amrex::ParallelFor( np, fillBufferFlag(pti, bmasks, inexflag, Geom(lev), 0) );
+    // - Find the indices that reorder particles so that the last particles
+    //   are in the larger buffer
+    fillWithConsecutiveIntegers( pid );
+    auto sep = stablePartition( pid.begin(), pid.end(), inexflag );
```
Comment (ax3l) [inline Source/Particles/Sorting/Partition.cpp]:
```suggestion
    auto const sep = stablePartition( pid.begin(), pid.end(), inexflag );
```

File: Source/Particles/Sorting/SortingUtils.H [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+#endif
+}
+
+/* \brief Find the indices that would reorder the elements of `predicate`
+ * so that the elements with non-zero value precede the other elements
+ *
+ * \param[in] index_begin Point to the beginning of the vector which is
```
Comment (ax3l) [inline Source/Particles/Sorting/SortingUtils.H]: I guess these are manipulated, so it's `[in,out]`? Same with the second argument.
I think the third argument is potentially a `const` reference to a `int const`?

File: Source/Particles/Sorting/SortingUtils.H [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+ * \param[in] Vector that indicates the elements that need to be reordered first
+ */
+template< typename ForwardIterator >
+ForwardIterator stablePartition(ForwardIterator index_begin,
+                               ForwardIterator index_end,
+                               amrex::Gpu::DeviceVector<int>& predicate)
+{
+#ifdef AMREX_USE_GPU
+    // On GPU: Use thrust
+    int* AMREX_RESTRICT predicate_ptr = predicate.dataPtr();
```
Comment (ax3l) [inline Source/Particles/Sorting/SortingUtils.H]: potentially a 
```suggestion
    int const * AMREX_RESTRICT predicate_ptr = predicate.dataPtr();
```

File: Source/Particles/Sorting/SortingUtils.H [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+    ForwardIterator sep = thrust::stable_partition(
+        thrust::cuda::par(amrex::Cuda::The_ThrustCachedAllocator()),
+        index_begin, index_end,
+        [predicate_ptr] AMREX_GPU_DEVICE (long i) { return predicate_ptr[i]; }
+    );
+#else
+    // On CPU: Use std library
+    ForwardIterator sep = std::stable_partition(
```
Comment (ax3l) [inline Source/Particles/Sorting/SortingUtils.H]:
```suggestion
    ForwardIterator const sep = std::stable_partition(
```

File: Source/Particles/Sorting/SortingUtils.H [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+ */
+template< typename ForwardIterator >
+ForwardIterator stablePartition(ForwardIterator index_begin,
+                               ForwardIterator index_end,
+                               amrex::Gpu::DeviceVector<int>& predicate)
+{
+#ifdef AMREX_USE_GPU
+    // On GPU: Use thrust
+    int* AMREX_RESTRICT predicate_ptr = predicate.dataPtr();
+    ForwardIterator sep = thrust::stable_partition(
```
Comment (ax3l) [inline Source/Particles/Sorting/SortingUtils.H]:
```suggestion
    ForwardIterator const sep = thrust::stable_partition(
```

File: Source/Particles/Sorting/SortingUtils.H [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+}
+
+/* \brief Return the number of elements between `first` and `last`
+ *
+ * \param[in] fist Points to a position in a vector
+ * \param[in] last Points to another position in a vector
+ * \return The number of elements between `first` and `last`
+ */
+template< typename ForwardIterator >
+int iteratorDistance(ForwardIterator first,
```
Comment (ax3l) [inline Source/Particles/Sorting/SortingUtils.H]: both arguments are `const`, I guess.

File: Source/Particles/Sorting/SortingUtils.H [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+                     ForwardIterator last)
+{
+#ifdef AMREX_USE_GPU
+    // On GPU: Use thrust
+    return thrust::distance( first, last );
```
Comment (ax3l) [inline Source/Particles/Sorting/SortingUtils.H]: If we use `thrust::distance` directly, we should include `<thrust/distance.h>` on `#ifdef AMREX_USE_GPU`

File: Source/Particles/Sorting/SortingUtils.H [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+ * \param[in] bmasks Spatial array, that contains a flag indicating
+ *         whether each cell is part of the gathering/deposition buffers
+ * \param[out] inexflag Vector to be filled with the value of `bmasks`
+ * \param[in] geom Geometry object, necessary to locate particles within the array `bmasks`
+ * \param[in] start_index Index that which elements start to be modified
+ */
+class fillBufferFlag
+{
+    public:
+        fillBufferFlag( WarpXParIter& pti, const amrex::iMultiFab* bmasks,
```
Comment (ax3l) [inline Source/Particles/Sorting/SortingUtils.H]: is `pti` also a `const` reference, according to `[in]`?
```suggestion
        fillBufferFlag( WarpXParIter const & pti, const amrex::iMultiFab* bmasks,
```

File: Source/Particles/Sorting/SortingUtils.H [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+                        amrex::Gpu::DeviceVector<int>& inexflag,
+                        const amrex::Geometry& geom, long start_index=0 ) {
```
Comment (ax3l) [inline Source/Particles/Sorting/SortingUtils.H]:
```suggestion
                        const amrex::Geometry& geom, long const start_index=0 ) {
```

File: Source/Particles/Sorting/SortingUtils.H [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+        int* m_inexflag_ptr;
+        WarpXParticleContainer::ParticleType* m_particles;
+        amrex::Array4<const int> m_buffer_mask;
+        long m_start_index;
+};
+
+/* \brief Functor that copies the elements of `src` into `dst`,
+ *       while reordering them according to `indices`
+ *
+ * \param src Source vector
```
Comment (ax3l) [inline Source/Particles/Sorting/SortingUtils.H]:
```suggestion
 * \param src[in] Source vector
```

File: Source/Particles/Sorting/SortingUtils.H [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+ * \param dst Destination vector
```
Comment (ax3l) [inline Source/Particles/Sorting/SortingUtils.H]:
```suggestion
 * \param dst[out] Destination vector
```

File: Source/Particles/Sorting/SortingUtils.H [hunk 16]
```diff
@@ -0,0 +1,164 @@
...
+ * \param indices Array of indices that indicate how to reorder elements
```
Comment (ax3l) [inline Source/Particles/Sorting/SortingUtils.H]: not sure:
```suggestion
 * \param indices[in,out] Array of indices that indicate how to reorder elements
```

File: Source/Particles/Sorting/SortingUtils.H [hunk 17]
```diff
@@ -0,0 +1,164 @@
...
+ */
+template <typename T>
+class copyAndReorder
+{
+    public:
+        copyAndReorder(
+            amrex::Gpu::ManagedDeviceVector<T>& src,
```
Comment (ax3l) [inline Source/Particles/Sorting/SortingUtils.H]: is `src` a `const` object?

File: Source/Particles/Sorting/SortingUtils.H [hunk 18]
```diff
@@ -0,0 +1,164 @@
...
+    ForwardIterator const sep = thrust::stable_partition(
+        thrust::cuda::par(amrex::Cuda::The_ThrustCachedAllocator()),
+        index_begin, index_end,
+        [predicate_ptr] AMREX_GPU_DEVICE (long i) { return predicate_ptr[i]; }
+    );
+#else
+    // On CPU: Use std library
+    ForwardIterator const sep = std::stable_partition(
+        index_begin, index_end,
+        [&predicate](long i) { return predicate[i]; }
```
Comment (ax3l) [inline Source/Particles/Sorting/SortingUtils.H]:
```suggestion
        [&predicate](long const i) { return predicate[i]; }
```

File: Source/Particles/Sorting/SortingUtils.H [hunk 19]
```diff
@@ -0,0 +1,164 @@
...
+                               ForwardIterator const index_end,
+                               amrex::Gpu::DeviceVector<int> const& predicate)
+{
+#ifdef AMREX_USE_GPU
+    // On GPU: Use thrust
+    int const* AMREX_RESTRICT predicate_ptr = predicate.dataPtr();
+    ForwardIterator const sep = thrust::stable_partition(
+        thrust::cuda::par(amrex::Cuda::The_ThrustCachedAllocator()),
+        index_begin, index_end,
+        [predicate_ptr] AMREX_GPU_DEVICE (long i) { return predicate_ptr[i]; }
```
Comment (ax3l) [inline Source/Particles/Sorting/SortingUtils.H]:
```suggestion
        [predicate_ptr] AMREX_GPU_DEVICE (long const i) { return predicate_ptr[i]; }
```

**Comment by ax3l** [comment]
@RemiLehe let us dig through a regression we saw after merging this. we have some holes on the main grid after merging this, which causes noise. maybe just an offset we miss somewhere.

**Pull request merged**